### PR TITLE
Riverpod 1.0.0

### PR DESCRIPTION
**Pull request by rrousselGit**
Implements #335 

TODO:

- [x] Update Consumer/ConsumerWidget
- [x] Remove useProvider
- [x] Add HookConsumer/HookConsumerWidget
- [x] add `ref.state` and `ref.state =` for providers
- [ ] refactor FutureProviderElement & co to proper classes instead of typedefs.
- [x] update all docs
- [x] implement ref.listen
- [x] implement ref.read (and deprecate context.read)
- [ ] ref.read should maintain the state of auto-dispose providers
- [x] support ref.watch(provider.select) inside providers
- [x] support ref.watch(provider.select) inside widgets
- [ ] remove ScopedProvider and allow all providers to be scoped.
- [ ] have FutureProvider/StreamProvider filter rebuilds with ==
- [x] pass the previous value to ProviderObserver.didUpdateProvider.
- [x] fixes https://github.com/rrousselGit/river_pod/issues/358
- [ ] allow `AsyncValue` to be subclassed
- [ ] have ref.listen receive the previous value, if any
- [ ] add tons of tests

### CODE REVIEW

**Comment by ResoDev** [comment]
Hello! I don't see anything like "deprecate `context.read(provider)`" in the checklist of this PR. Does that mean that `ref.read(provider)` won't be implemented?

**Comment by rrousselGit** [comment]
Oh I just forgot to add it to the list @ResoDev

**Comment by TimWhiting** [comment]
Will you also be deprecating context.refresh -> ref.refresh?

**Comment by rrousselGit** [comment]
Yes

And remove `ref.container` too

**Review by TimWhiting — COMMENTED** [review]
Changes in `riverpod_cli` should be reverted.

File: packages/flutter_riverpod/lib/src/consumer.dart [hunk 1]
```diff
@@ -4,12 +4,17 @@ import 'package:meta/meta.dart';
 import 'framework.dart';
 import 'internals.dart';
 
+// ignore: one_member_abstracts
```
Comment (kranfix) [inline packages/flutter_riverpod/lib/src/consumer.dart:7]: @rrousselGit I think there should exist a `ProviderReferenceBase`. Then `ProviderReference` and `WidgetReference` must extend it becase they will have a common API.

**Comment by SalahAdDin** [comment]
What's about the following?

> Avoid calling `context.read` inside the build method of a Widget.
> If you want to optimize rebuilds, extract the value listened in a Provider instead.

**Comment by rrousselGit** [comment]
> What's about the following?
> 
> > Avoid calling `context.read` inside the build method of a Widget.
> > If you want to optimize rebuilds, extract the value listened in a Provider instead.

That is going to stay.

**Comment by tbm98** [comment]
> > What's about the following?
> > > Avoid calling `context.read` inside the build method of a Widget.
> > > If you want to optimize rebuilds, extract the value listened in a Provider instead.
> 
> That is going to stay.

No. It should be modify if we have `ref.watch(provider.select)`

**Comment by rrousselGit** [comment]
I'm not following.

Why would select invalidate this warning?

**Comment by tbm98** [comment]
> Avoid calling context.read inside the build method of a Widget.

it's fine.

but 
>If you want to optimize rebuilds, extract the value listened in a Provider instead.

we can do with ref.watch(provider.select) without extract to new provider

**Comment by rrousselGit** [comment]
Oh you are correct. My bad

File: packages/hooks_riverpod/CHANGELOG.md [hunk 2]
```diff
@@ -197,7 +197,7 @@ Fixed a bug that prevented the use of `ConsumerWidget` under normal circumstance
     const MyWidget({Key? key}) : super(key: key);
 
     @override
-    Widget build(BuildContext context, ScopedReader watch) {
+    Widget build(BuildContext context, WidgetReference ref) {
       final value = watch(myProvider);
```
Comment (kranfix) [inline packages/hooks_riverpod/CHANGELOG.md:201]: This line must be `final value = ref.watch(myProvider);`

**Review by kranfix — CHANGES_REQUESTED** [review]

**Comment by TimWhiting** [comment]
#530 should fix CI for this branch

**Comment by rrousselGit** [comment]
Let's merge this, make a dev release, and do the remaining tasks in a separate PR

**Pull request merged**